### PR TITLE
Support MERGE on distributed tables with restrictions

### PR DESCRIPTION
**Pull request by tejeswarm**
This implements the phase - II of MERGE sql support

Support routable query where all the tables in the merge-sql are distributed, co-located, and both the source and target relations are joined on the distribution column with a constant qual. This should be a Citus single-task query. Below is an example.

SELECT create_distributed_table('t1', 'id'); 
SELECT create_distributed_table('s1', 'id', colocate_with => ‘t1’); 

MERGE INTO t1 
USING s1 ON **t1.id = s1.id** AND **t1.id = 100** 
      		WHEN MATCHED THEN 
UPDATE SET val = s1.val + 10 
       		WHEN MATCHED THEN 
DELETE 
WHEN NOT MATCHED THEN 
INSERT (id, val, src) VALUES (s1.id, s1.val, s1.src) 

Basically, MERGE checks to see if
 - There are a minimum of two distributed tables (source and a target).
 - All the distributed tables are indeed colocated.
 - MERGE relations are joined on the distribution column
          MERGE .. USING .. ON target.dist_key = source.dist_key
 - The query should touch only a single shard i.e. JOIN AND with a constant qual
       MERGE .. USING .. ON target.dist_key = source.dist_key AND target.dist_key = <>

If any of the conditions are not met, it raises an exception.

### CODE REVIEW

File: src/backend/distributed/planner/multi_router_planner.c [hunk 1]
```diff
@@ -620,148 +779,45 @@ ModifyPartialQuerySupported(Query *queryTree, bool multiShardQuery,
...
-	Var *partitionColumn = NULL;
-	if (IsCitusTable(resultRelationId))
-	{
-		partitionColumn = PartitionColumn(resultRelationId, rangeTableId);
-	}
 	commandType = queryTree->commandType;
 	if (commandType == CMD_INSERT || commandType == CMD_UPDATE ||
 		commandType == CMD_DELETE)
 	{
-		bool hasVarArgument = false; /* A STABLE function is passed a Var argument */
```
Comment (tejeswarm) [inline src/backend/distributed/planner/multi_router_planner.c:634]: This code is simply moved to a new routine

File: src/backend/distributed/planner/distributed_planner.c [hunk 2]
```diff
@@ -2692,43 +2798,69 @@ ErrorIfMergeHasUnsupportedTables(Query *parse, List *rangeTableList)
...
+			ereport(ERROR, (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+							errmsg("MERGE command is not supported with "
+								   "combination of distributed/local tables yet")));
+		}
+		else
+		{
+			/* All the tables are local, supported */
+			return;
+		}
+	}
```
Comment (aykut-bozkurt) [inline src/backend/distributed/planner/distributed_planner.c:2862]:
```suggestion
	/* Ensure all tables are indeed local */
	if (foundLocalTables && list_length(distTablesList) == 0)
	{
		/* All the tables are local, supported */
		return;
	}
	else if (foundLocalTables && list_length(distTablesList) > 0)
	{
		ereport(ERROR, (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
						errmsg("MERGE command is not supported with "
								"combination of distributed/local tables yet")));
	}
```

Maybe we can get rid of 1 level of indenting as it seems easier to follow.

Comment (tejeswarm) [reply src/backend/distributed/planner/distributed_planner.c:2862]: Done.

File: src/backend/distributed/planner/distributed_planner.c [hunk 3]
```diff
@@ -2614,15 +2626,109 @@ IsMergeAllowedOnRelation(Query *parse, RangeTblEntry *rte)
...
+		}
+
+		if (currentColocationId != savedColocationId)
+		{
+			/* All distributed tables must be colocated */
+			ereport(ERROR, (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+							errmsg("For MERGE command, all the distributed tables "
+								   "must be colocated")));
+		}
+	}
```
[the remark this answers is not in the document]
Comment (tejeswarm) [reply src/backend/distributed/planner/distributed_planner.c:2686]: Yes, I did check that routine, couple of issues
1) Minor (and easily fixable): It didn't specifically check for range (append is there) and raise a customized error message for these types of tables, but with a return value of colocation in a pointer-parameter we can get this done.
2) Major: The routine walks through the range table list, `RelationRestriction`, from the planner and in this we are using range table list of `RangeTblEntry` from parser (it walks through the query tree and prepares it in `ExtractRangeTableEntryList`)

But having said that, by this point, we have already eliminated all the local and reference tables (if any), so, the list must match I believe, but we need to do more diligence in establishing that.

**Comment by tejeswarm** [comment]
> behaviour around subqueries is a bit confusing:
> 
> ```
> MERGE INTO dist_target
> USING (SELECT * FROM dist_colocated WHERE id = 1) sub
> ON dist_target.id = sub.id WHEN MATCHED THEN
> UPDATE SET val = sub.val
> WHEN NOT MATCHED THEN
> INSERT VALUES(sub.id, sub.val);
> ERROR:  MERGE command must be joined with a constant qual on target's distribution column
> ```

-- sub-query as a source
BEGIN;
MERGE INTO target_cj t
USING (SELECT * FROM source_cj1 WHERE sid1 = 2) sub
ON t.tid = sub.sid1
WHEN MATCHED THEN
        UPDATE SET src = sub.src1
WHEN NOT MATCHED THEN
        DO NOTHING;
ERROR:  MERGE command must be joined with a constant qual on target's distribution column
SELECT * FROM target_cj ORDER BY 1;
ERROR:  current transaction is aborted, commands ignored until end of transaction block
ROLLBACK;
--

**Comment by tejeswarm** [comment]
> behaviour around subqueries is a bit confusing:
> 
> ```
> MERGE INTO dist_target
> USING (SELECT * FROM dist_colocated WHERE id = 1) sub
> ON dist_target.id = sub.id WHEN MATCHED THEN
> UPDATE SET val = sub.val
> WHEN NOT MATCHED THEN
> INSERT VALUES(sub.id, sub.val);
> ERROR:  MERGE command must be joined with a constant qual on target's distribution column
> ```

That's expected behavior for phase-II, we have to explicitly make it a single shard task by adding a condition, something like **AND t.tid = 2**

- The query should touch only a single shard i.e. JOIN AND with a constant qual
           **MERGE .. USING .. ON target.dist_key = source.dist_key AND target.dist_key = <>**


```
-- sub-query as a source
BEGIN;
MERGE INTO target_cj t
USING (SELECT * FROM source_cj1 WHERE sid1 = 2) sub
ON t.tid = sub.sid1 AND t.tid = 2
WHEN MATCHED THEN
        UPDATE SET src = sub.src1
WHEN NOT MATCHED THEN
        DO NOTHING;
SELECT * FROM target_cj ORDER BY 1;
 tid |   src    | val
---------------------------------------------------------------------
   1 | target   |   0
   2 | source-1 |   0
   2 | source-1 |   0
   3 | target   |   0
(4 rows)

ROLLBACK;
```

**Comment by marcocitus** [comment]
I mainly meant that the error message is confusing, because we are joining with the distribution column.

**Comment by tejeswarm** [comment]
> * ON target.dist_key = source.dist_key AND target.dist_key = <>

 ```
errmsg("MERGE command's ON clause must have a constant "
                                                           "qual on target's distribution column"),
                                                errhint("Consider adding .. AND target.dist_key = <>")));
```

**Comment by tejeswarm** [comment]
> I mainly meant that the error message is confusing, because we are joining with the distribution column.

Does this look better?
```
errmsg("MERGE command's ON clause must have a constant "
               "qual on target's distribution column"),
errhint("Consider adding .. AND target.dist_key = <>")));

```

File: src/backend/distributed/planner/distributed_planner.c [hunk 4]
```diff
@@ -2614,19 +2627,112 @@ IsMergeAllowedOnRelation(Query *parse, RangeTblEntry *rte)
...
+	{
+		quals = (Node *) make_ands_explicit((List *) quals);
+	}
+
+	if (!ConjunctionContainsColumnFilter(quals, distributionKey, &distributionKeyValue))
+	{
+		ereport(ERROR, (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+						errmsg("MERGE command's ON clause must have a constant "
+							   "qual on target's distribution column"),
+						errhint("Consider adding .. AND target.dist_key = <>")));
```
Comment (marcocitus) [inline src/backend/distributed/planner/distributed_planner.c:2716]: minor, maybe:
```
						errmsg("MERGE on a distributed table requires a constant filter "
                               "on the distribution column of the target table"),
						errhint("Consider adding .. AND target.dist_key = <> to the ON clause")));
```

**Comment by marcocitus** [comment]
I was surprised this (self-join) query failed:
```
MERGE INTO dist_colocated dist_target
USING (SELECT * FROM dist_colocated) sub
ON dist_target.id = sub.id AND dist_target.id = 1 WHEN MATCHED THEN
UPDATE SET val = sub.val
WHEN NOT MATCHED THEN
INSERT VALUES(sub.id, sub.val);
ERROR:  MERGE currently supports single shard queries only, this command is accessing (32) shards
```

when this one succeeds:
```
MERGE INTO dist_target
USING (SELECT * FROM dist_colocated) sub
ON dist_target.id = sub.id AND dist_target.id = 1 WHEN MATCHED THEN
UPDATE SET val = sub.val
WHEN NOT MATCHED THEN
INSERT VALUES(sub.id, sub.val);
```
(only difference is first line)

**Review by marcocitus — COMMENTED** [review]
Overall the code looks pretty reasonable to me, so it seems ready after resolving a few small(?) issues and adding tests for those.

File: src/backend/distributed/planner/distributed_planner.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (tejeswarm) [reply src/backend/distributed/planner/distributed_planner.c]: Made an attempt to converge both, not the cleanest way, but works. Please take a look.

**Comment by tejeswarm** [comment]
> I was surprised this (self-join) query failed:
> 
> ```
> MERGE INTO dist_colocated dist_target
> USING (SELECT * FROM dist_colocated) sub
> ON dist_target.id = sub.id AND dist_target.id = 1 WHEN MATCHED THEN
> UPDATE SET val = sub.val
> WHEN NOT MATCHED THEN
> INSERT VALUES(sub.id, sub.val);
> ERROR:  MERGE currently supports single shard queries only, this command is accessing (32) shards
> ```
> 
> when this one succeeds:
> 
> ```
> MERGE INTO dist_target
> USING (SELECT * FROM dist_colocated) sub
> ON dist_target.id = sub.id AND dist_target.id = 1 WHEN MATCHED THEN
> UPDATE SET val = sub.val
> WHEN NOT MATCHED THEN
> INSERT VALUES(sub.id, sub.val);
> ```
> 
> (only difference is first line)

The former one is a self-join (on target) and the latter is a regular join of two different tables. Fixed the former case (good catch!!!)

**Comment by tejeswarm** [comment]
> Issue with prepared statements:
> 
> ```
> PREPARE foo(int) AS MERGE INTO dist_target dist_target
> USING (SELECT * FROM dist_colocated) sub
> ON dist_target.id = sub.id AND dist_target.id = $1 WHEN MATCHED THEN
> UPDATE SET val = sub.val
> WHEN NOT MATCHED THEN
> INSERT VALUES(sub.id, sub.val);
> EXECUTE foo(1);
> EXECUTE foo(1);
> EXECUTE foo(1);
> EXECUTE foo(1);
> EXECUTE foo(1);
> EXECUTE foo(1);
> ERROR:  MERGE currently supports single shard queries only, this command is accessing (32) shards
> ```

This is a bit of tricky (at least in phase-II as we are enforcing single-shard requirement). As you already know, the first five plans are **custom plans** i.e., when planning we see filter clause as
` {type = T_Var} = {type = T_Const}` , which prunes the shard-list to 1 for target table, which is what we desire.

On sixth attempt we are generating a **generic plan** with
`{type = T_Var} = {type = T_Param}`, which is not pruning the shards for target table and is raising the exception.

I tried to replicate this in a simple non-merge sql join

PREPARE foo(int) AS SELECT * FROM target **left join** source on target.tid=source.sid where tid=$1;

But this one is always pruning the shards-list to 1, because even after sixth iteration we are **always** choosing a **custom pla**n as the cost of generic plan (generic_cost = 3.4028233462973677e+30) is way higher than the custom plan (avg_custom_cost = 12.5)
 
As an experiment, I set a high generic cost value for MERGE to plansource->generic_cost = 100000, and it all works fine as we are repeatedly choosing the custom plans. 
Edit: Even this option `set plan_cache_mode = force_custom_plan;` works. I've added both postive and negative test cases. As the single shard requirement is only for this phase-II, please share your input on if we should invest in making prepare work for merge or reassess it in the next phase.

**Comment by marcocitus** [comment]
>  please share your input on if we should invest in making prepare work for merge or reassess it in the next phase.

I don't think there's a point in adding the feature if we error out for prepared statements, since that's how many drivers do their queries.

File: src/backend/distributed/planner/distributed_planner.c [hunk 5]
```diff
@@ -261,6 +264,15 @@ distributed_planner(Query *parse,
 												planContext.boundParams);
 			if (needsDistributedPlanning)
 			{
+				/*
+				 * Fast path queries cannot have merge command, and we
+				 * prevent the remaining here.
+				 */
+				ErrorIfQueryHasUnsupportedMergeCommand(planContext.originalQuery,
```
Comment (marcocitus) [inline src/backend/distributed/planner/distributed_planner.c:271]: I kind of missed that we added this line to this super important top-level function.

Can we move this to a more appropriate place? (e.g. router planner, and make other planners error for CMD_MERGE)

(the below call to TryToDelegateFunctionCall is already a quite ugly piece of this function, but at least it has some rationale behind it since it concerns a query that has no distributed tables)

Comment (tejeswarm) [reply src/backend/distributed/planner/distributed_planner.c]: OK, we simply piggybacked on an existing check that was added as part of PG-15 support. In this phase, we are doing some of the deeper checks, such as updating-distribution-column further down in `ModifyQuerySupported->MergeQuerySupported`, it makes sense to converge them.

**Comment by tejeswarm** [comment]
> > please share your input on if we should invest in making prepare work for merge or reassess it in the next phase.
> 
> I don't think we should add the feature if we error out for prepared statements, since that's how many drivers do their queries. The cost trick has so far worked reasonably well for the cases where we don't do any meaningful caching.

Yup, dissuading the planner worked well, removed the negative test for prepared statements, all good now, please check.

**Comment by tejeswarm** [comment]
> > > please share your input on if we should invest in making prepare work for merge or reassess it in the next phase.
> > 
> > 
> > I don't think we should add the feature if we error out for prepared statements, since that's how many drivers do their queries. The cost trick has so far worked reasonably well for the cases where we don't do any meaningful caching.
> 
> Yup, dissuading the planner worked well, removed the negative test for prepared statements, all good now, please check.

File: src/backend/distributed/planner/multi_physical_planner.c [hunk 6]
```diff
@@ -2224,17 +2228,8 @@ QueryPushdownSqlTaskList(Query *query, uint64 jobId,
...
-		 * all tables of the left hand side. However if the right hand side is bigger than
-		 * the left hand side we don't have to send the query to any shard that is not
-		 * matching anything on the left hand side.
-		 *
-		 * Instead we will simply skip any RelationRestriction if it is an OUTER join and
-		 * the table is part of the non-outer side of the join.
-		 */
-		if (IsInnerTableOfOuterJoin(relationRestriction))
+		/* Check to see if we can skip this relation */
+		if (SkipRelationInJoin(query, relationRestriction, prunedShardList))
```
[the remark this answers is not in the document]
Comment (tejeswarm) [reply src/backend/distributed/planner/multi_physical_planner.c]: As we are not restricting shards for source (with no constant filter), we still generate multi-shards router query on it. We need the query routed to a single shard identified by the target constant filter. Removed the function and in-lined the code.

File: src/backend/distributed/planner/multi_physical_planner.c [hunk 7]
```diff
@@ -2302,6 +2318,45 @@ QueryPushdownSqlTaskList(Query *query, uint64 jobId,
...
+ * IsOuterTableOfOuterJoin tests based on the join information envoded in a
+ * RelationRestriction if the table accessed for this relation is
+ *   a) in an outer join
+ *   b) on the outer part of said join
+ *
+ * The function returns true only if both conditions above hold true
+ */
+static bool
+IsOuterTableOfOuterJoin(RelationRestriction *relationRestriction)
+{
```
Comment (marcocitus) [inline src/backend/distributed/planner/multi_physical_planner.c:2331]: can you double check the code coverage? might just be a blip

**Review by marcocitus — APPROVED** [review]

File: src/backend/distributed/planner/multi_physical_planner.c (hunk 7 above)
Comment (tejeswarm) [reply src/backend/distributed/planner/multi_physical_planner.c:2331]: Sure, will check before merging.

Comment (tejeswarm) [reply src/backend/distributed/planner/multi_physical_planner.c:2331]: Ran code-cov couple of times, it's all good now.
